damnit/util.py

```diff
@@ -14,3 +23,11 @@ def wait_until(condition, timeout=1):
 
     if slept_for >= timeout:
         raise TimeoutError("Condition timed out")
+
+def timestamp2str(timestamp):
+    if timestamp is None or pd.isna(timestamp):
+        return None
+
+    dt_naive = datetime.fromtimestamp(timestamp)
+    dt_local = dt_naive.replace(tzinfo=timezone.utc).astimezone()
```

Comment:
```suggestion
    dt_utc = datetime.fromtimestamp(timestamp, timezone.utc)
    dt_local = dt_utc.astimezone()
```

`fromtimestamp()` without a timezone actually gives you a naive datetime representing *local* time, which `.replace()` then labels as UTC, so you double the offset. Compare:

```
In [12]: datetime.fromtimestamp(ts).replace(tzinfo=timezone.utc).astimezone().strftime("%H:%M:%S %d/%m/%Y")
Out[12]: '19:05:24 07/07/2023'

In [13]: datetime.fromtimestamp(ts, timezone.utc).astimezone().strftime("%H:%M:%S %d/%m/%Y")
Out[13]: '18:05:24 07/07/2023'
```

My local clock was 18:05 when I did `ts = time.time()`, and that's 17:05 in UTC. The first version puts me 2 hours ahead instead of 1.

You could also just leave it with the naive local datetime and format that:

```python
dt_local = datetime.fromtimestamp(timestamp)
return dt_local.strftime("%H:%M:%S %d/%m/%Y")
```

But I like to make datetimes timezone-aware wherever possible. :slightly_smiling_face: 
```

Reply:
You're quite right, but now I'm massively confused because the current code is supposedly giving the right timestamps :upside_down_face: Taking p3303 as an example, with the current code run 1 is supposed to have started at 1:14PM:
![image](https://github.com/European-XFEL/DAMNIT/assets/5361518/948a715b-f94f-4b7c-b866-27084539a729)

Which according to [mymdc](https://in.xfel.eu/metadata/proposals/819/runs/176414) is correct:
![image](https://github.com/European-XFEL/DAMNIT/assets/5361518/b2aae9d6-b8c2-43f6-b4e2-e1f452af417f)

If I try with your code I get a start time of 12:14PM instead, an hour earlier. So there's definitely an extra hour being added, but does that mean that the time in mymdc is incorrect? I can't see any other reason for the discrepancy, your explanation makes sense to me.

Reply:
Oh DAMNIT :tm: . There's a bug with how we're getting the timestamp to store in the database. You see the correct answer because the two bugs cancel each other out. Timezones! :exploding_head: 

https://github.com/European-XFEL/DAMNIT/blob/d5e5315f873bb27662b65525ca186d233e70533d/damnit/ctxsupport/ctxrunner.py#L184-L185

The Numpy datetime64 object implicitly represents time in UTC, but it doesn't store any timezone info. `.item()` converts it to a naive Python `datetime` object, which `.timestamp()` then treats as local time ([as documented](https://docs.python.org/3/library/datetime.html#datetime.datetime.timestamp)). So we're subtracting 1 or 2 hours (winter/summer) before saving it in the database.

We should be doing something like:

```python
return np.datetime64(ts, 'us').item().replace(tzinfo=timezone.utc).timestamp()
```

I'm inclined to fix this, even if it means existing databases are left showing the wrong start times in the GUI. :disappointed:

Reply:
I've opened #105 to fix the extraction code.

Reply:
Fixed in the GUI in 2931215, tested by checking run 290 from p4239 :clock12: